### PR TITLE
Added target options pass-through

### PR DESCRIPTION
**Pull request by josh146**
**Context:** Currently, target options defined in a loaded Blackbird script are passed to `program.compile`, where they are ignored.

**Description of the Change:**

* `Program` now contains the attribute `Program.run_options`, storing any default run options for a program. During compilation, target options from the blackbird scripts are now stored in this attribute.

* During `eng.run`, if no run options are provided as a keyword argument, the engine will fall back on the run options stored within the program.

* Otherwise, if run options are provided by the user, they 'overwrite' the program's default run options.

**Benefits:**

* Blackbird `target (shots=100)` directives are now stored in the compiled SF Program, and used as the default run options in `eng.run()`. The user may still overwrite these default by passing in `eng.run(prog, run_options={"shots": 24})`.

**Possible Drawbacks:** None that I can think of.

**Related GitHub Issues:** #129

### CODE REVIEW

File: strawberryfields/program.py [hunk 1]
```diff
@@ -552,6 +559,12 @@ def compile_sequence(seq):
         compiled = self._linked_copy()
         compiled.circuit = seq
         compiled.target = target
+
+        # get run options of compiled program
+        # for the moment, shots is the only supported run option.
+        if "shots" in kwargs:
+            compiled.run_options["shots"] = kwargs["shots"]
```
Comment (josh146) [inline strawberryfields/program.py]: any others we should add here?

File: tests/frontend/test_io.py [hunk 2]
```diff
@@ -453,6 +459,65 @@ def test_invalid_compilation(self):
...
+    """A dummy run function, that when called returns a dummy
+    results object, with run options available as an attribute.
+    This allows run_options to be returned and inspected after calling eng.run
+    """
+    results = DummyResults()
+    results.run_options = eng_run_options
+    return results
+
+
+class TestEngineIntegration:
```
Comment (co9olguy) [inline tests/frontend/test_io.py]: I guess this is a "frontend integration" test?

File: strawberryfields/engine.py [hunk 3]
```diff
@@ -367,21 +367,34 @@ def run(self, program, *, compile_options=None, run_options=None):
                 and state. TF backend only.
         """
         compile_options = compile_options or {}
-        run_options = run_options or {}
-        run_options.setdefault("shots", 1)
-        run_options.setdefault('modes', None)
+        temp_run_options = {}
+
+        if isinstance(program, Sequence):
+            # succesively update all run option defaults.
```
Comment (co9olguy) [inline strawberryfields/engine.py]: Feels like this is something that warrants a dedicated test

File: strawberryfields/program.py (hunk 1 above)
Comment (co9olguy) [reply strawberryfields/program.py]: for now, probably not

File: tests/frontend/test_io.py (hunk 2 above)
Comment (josh146) [reply tests/frontend/test_io.py]: I guess so. Frontend integration in that it is testing multiple parts of the frontend (io.load, Program, eng.run()). But everything is still contained within the frontend.

File: strawberryfields/engine.py (hunk 3 above)
Comment (josh146) [reply strawberryfields/engine.py]: Will add.

Comment (josh146) [reply strawberryfields/engine.py]: Added in 8fe10dd2eb61bfb02b4c20942f347e8aa3493ca6

**Pull request merged**